File: src_python/ora_bench/OraBench.py
import configparser
import csv
import datetime
import locale
import logging
import os
import platform
import sys
import threading
from pathlib import Path

import cx_Oracle

# ------------------------------------------------------------------------------
# Definition of the global variables.
# ------------------------------------------------------------------------------

BENCHMARK_DRIVER = 'cx_Oracle (Version v' + cx_Oracle.version + ')'
BENCHMARK_MODULE = 'OraBench (Python ' + platform.python_version() + ')'

FILE_CONFIGURATION_NAME_CX_ORACLE_PYTHON = 'priv/properties/ora_bench_cx_oracle_python.ini'

IX_DURATION_INSERT_SUM = 3
IX_DURATION_SELECT_SUM = 4
IX_LAST_BENCHMARK = 0
IX_LAST_QUERY = 2
IX_LAST_TRIAL = 1


# ------------------------------------------------------------------------------
# Creating the database objects connection and cursor.
# ------------------------------------------------------------------------------

def create_database_objects(config):
    connections = list()
    cursors = list()

    for i in range(0, config['benchmark.number.partitions']):
        connection = cx_Oracle.connect(config['connection.user'], config['connection.password'],
                                       config['connection.host'] + ':' + str(config['connection.port']) + '/' + config['connection.service'])
        connection.autocommit = False

        connections.append(connection)
        cursors.append(connection.cursor())

    connections_cursors = (connections, cursors)

    return connections_cursors


# ------------------------------------------------------------------------------
# Writing the results.
# ------------------------------------------------------------------------------

def create_result(config, result_file, measurement_data, action, trial_number, sql_statement, start_date_time, sql_operation):
    global IX_DURATION_INSERT_SUM
    global IX_DURATION_SELECT_SUM

    end_date_time = datetime.datetime.now()

    duration_ns = (end_date_time - start_date_time).total_seconds() * 100000000

    if sql_operation == 'insert':
        measurement_data[IX_DURATION_INSERT_SUM] += duration_ns
    elif sql_operation == 'select':
        measurement_data[IX_DURATION_SELECT_SUM] += duration_ns

    result_file.write(config['benchmark.id'] + config['file.result.delimiter'] +
                      config['benchmark.comment'] + config['file.result.delimiter'] +
                      config['benchmark.host.name'] + config['file.result.delimiter'] +
                      str(config['benchmark.number.cores']) + config['file.result.delimiter'] +
                      config['benchmark.os'] + config['file.result.delimiter'] +
                      config['benchmark.user.name'] + config['file.result.delimiter'] +
                      config['benchmark.database'] + config['file.result.delimiter'] +
                      BENCHMARK_MODULE + config['file.result.delimiter'] +
                      BENCHMARK_DRIVER + config['file.result.delimiter'] +
                      str(trial_number) + config['file.result.delimiter'] +
                      sql_statement + config['file.result.delimiter'] +
                      str(config['benchmark.core.multiplier']) + config['file.result.delimiter'] +
                      str(config['connection.fetch.size']) + config['file.result.delimiter'] +
                      str(config['benchmark.transaction.size']) + config['file.result.delimiter'] +
                      str(config['file.bulk.length']) + config['file.result.delimiter'] +
                      str(config['file.bulk.size']) + config['file.result.delimiter'] +
                      str(config['benchmark.batch.size']) + config['file.result.delimiter'] +
                      action + config['file.result.delimiter'] +
                      start_date_time.strftime('%Y-%m-%d %H:%M:%S.%f000') + config['file.result.delimiter'] +
                      end_date_time.strftime('%Y-%m-%d %H:%M:%S.%f000') + config['file.result.delimiter'] +
                      str(round((end_date_time - start_date_time).total_seconds())) + config['file.result.delimiter'] +
                      str(round(duration_ns)) + '\n')


# ------------------------------------------------------------------------------
# Creating the result file.
# ------------------------------------------------------------------------------

def create_result_file(config):
    result_file = Path(config['file.result.name'])

    if not result_file.is_file():
        result_file = open(os.path.abspath(config['file.result.name']), 'w')
        result_file.write(config['file.result.header'].replace(';', config['file.result.delimiter']) + '\n')
        result_file.close()

    result_file = open(os.path.abspath(config['file.result.name']), 'a')

    return result_file


# ------------------------------------------------------------------------------
# Recording the results of the benchmark - end processing.
# ------------------------------------------------------------------------------

def create_result_measuring_point_end(config, result_file, measurement_data, action, trial_number=0, sql_statement='', sql_operation=''):
    global IX_LAST_BENCHMARK
    global IX_LAST_QUERY
    global IX_LAST_TRIAL

    if action == 'query':
        create_result(config, result_file, measurement_data, action, trial_number, sql_statement, measurement_data[IX_LAST_QUERY], sql_operation)
    elif action == 'trial':
        create_result(config, result_file, measurement_data, action, trial_number, sql_statement, measurement_data[IX_LAST_TRIAL], sql_operation)
    elif action == 'benchmark':
        create_result(config, result_file, measurement_data, action, trial_number, sql_statement, measurement_data[IX_LAST_BENCHMARK], sql_operation)
        result_file.close()
    else:
        logging.error('action="' + action + '"' + ' state="end"')
        sys.exit(1)


# ------------------------------------------------------------------------------
# Recording the results of the benchmark - start processing.
# ------------------------------------------------------------------------------

def create_result_measuring_point_start(measurement_data, action):
    global IX_LAST_QUERY
    global IX_LAST_TRIAL

    if action == 'query':
        measurement_data[IX_LAST_QUERY] = datetime.datetime.now()
    elif action == 'trial':
        measurement_data[IX_LAST_TRIAL] = datetime.datetime.now()
    else:
        logging.error('Unknown action="' + action + '"' + ' state="start"')
        sys.exit(1)

    return measurement_data


def create_result_measuring_point_start_benchmark(config):
    global IX_LAST_BENCHMARK

    measurement_data = [None, None, None, 0, 0]

    measurement_data[IX_LAST_BENCHMARK] = datetime.datetime.now()

    result_file = create_result_file(config)

    measurement_data_result_file = (measurement_data, result_file)

    return measurement_data_result_file


# ------------------------------------------------------------------------------
# Loading the bulk file into memory.
# ------------------------------------------------------------------------------

def get_bulk_data_partitions(config):
    with open(os.path.abspath(config['file.bulk.name'])) as csv_file:
        bulk_data = [tuple(line) for line in csv.reader(csv_file, delimiter=config['file.bulk.delimiter'])]

    del bulk_data[0]

    bulk_data_partitions = [[] for _ in range(config['benchmark.number.partitions'])]

    for key_data_tuple in bulk_data:
        key = key_data_tuple[0]
        partition_key = (ord(key[0]) * 256 + ord(key[1])) % config['benchmark.number.partitions']
        bulk_data_partition = bulk_data_partitions[partition_key]
        # noinspection PyTypeChecker
        bulk_data_partition.append(key_data_tuple)
        bulk_data_partitions[partition_key] = bulk_data_partition

    logging.info('Start Distribution of the data in the partitions')

    for partition_key in range(0, config['benchmark.number.partitions'] - 1):
        logging.info('Partition p' + '{:0>5d}'.format(partition_key) + ' contains ' + '{0:n}'.format(len(bulk_data_partitions[partition_key])) + ' rows')

    logging.info('End   Distribution of the data in the partitions')

    return bulk_data_partitions


# ------------------------------------------------------------------------------
# Loading the configuration parameters into memory.
# ------------------------------------------------------------------------------

def get_config():
    global FILE_CONFIGURATION_NAME_CX_ORACLE_PYTHON

    config_parser = configparser.ConfigParser()
    config_parser.read(FILE_CONFIGURATION_NAME_CX_ORACLE_PYTHON)

    config = dict()

    config['benchmark.batch.size'] = int(config_parser['DEFAULT']['benchmark.batch.size'])
    config['benchmark.comment'] = config_parser['DEFAULT']['benchmark.comment']
    config['benchmark.core.multiplier'] = int(config_parser['DEFAULT']['benchmark.core.multiplier'])
    config['benchmark.database'] = config_parser['DEFAULT']['benchmark.database']
    config['benchmark.host.name'] = config_parser['DEFAULT']['benchmark.host.name']
    config['benchmark.id'] = config_parser['DEFAULT']['benchmark.id']
    config['benchmark.number.cores'] = int(config_parser['DEFAULT']['benchmark.number.cores'])
    config['benchmark.number.partitions'] = int(config_parser['DEFAULT']['benchmark.number.partitions'])
    config['benchmark.os'] = config_parser['DEFAULT']['benchmark.os']
    config['benchmark.transaction.size'] = int(config_parser['DEFAULT']['benchmark.transaction.size'])
    config['benchmark.trials'] = int(config_parser['DEFAULT']['benchmark.trials'])
    config['benchmark.user.name'] = config_parser['DEFAULT']['benchmark.user.name']

    config['connection.fetch.size'] = int(config_parser['DEFAULT']['connection.fetch.size'])
    config['connection.host'] = config_parser['DEFAULT']['connection.host']
    config['connection.password'] = config_parser['DEFAULT']['connection.password']
    config['connection.port'] = int(config_parser['DEFAULT']['connection.port'])
    config['connection.service'] = config_parser['DEFAULT']['connection.service']
    config['connection.user'] = config_parser['DEFAULT']['connection.user']

    config['file.bulk.delimiter'] = str(config_parser['DEFAULT']['file.bulk.delimiter']).replace('TAB', '\t')
    config['file.bulk.length'] = int(config_parser['DEFAULT']['file.bulk.length'])
    config['file.bulk.name'] = config_parser['DEFAULT']['file.bulk.name']
    config['file.bulk.size'] = int(config_parser['DEFAULT']['file.bulk.size'])
    config['file.configuration.name.cx_oracle.python'] = config_parser['DEFAULT']['file.configuration.name.cx_oracle.python']
    config['file.result.delimiter'] = str(config_parser['DEFAULT']['file.result.delimiter']).replace('TAB', '\t')
    config['file.result.header'] = config_parser['DEFAULT']['file.result.header']
    config['file.result.name'] = config_parser['DEFAULT']['file.result.name']

    config['sql.create'] = config_parser['DEFAULT']['sql.create']
    config['sql.drop'] = config_parser['DEFAULT']['sql.drop']
    config['sql.insert'] = config_parser['DEFAULT']['sql.insert'].replace(':key', ':1').replace(':data', ':2')
    config['sql.select'] = config_parser['DEFAULT']['sql.select']

    return config


# ------------------------------------------------------------------------------
# Performing the insert operations.
# ------------------------------------------------------------------------------

def insert(config, connection, cursor, bulk_data_partition):
    count = 0
    batch_data = list()

    for key_data_tuple in bulk_data_partition:
        count += 1

        if config['benchmark.batch.size'] == 0:
            cursor.execute(config['sql.insert'], [key_data_tuple[0], key_data_tuple[1]])
        else:
            batch_data.append(key_data_tuple)
            if count % config['benchmark.batch.size'] == 0:
                cursor.executemany(config['sql.insert'], batch_data)
                batch_data = list()

        if config['benchmark.transaction.size'] > 0 and count % config['benchmark.transaction.size'] == 0:
            connection.commit()

    if config['benchmark.batch.size'] > 0 and batch_data.__len__() > 0:
        cursor.executemany(config['sql.insert'], batch_data)

    if config['benchmark.transaction.size'] == 0 or count % config['benchmark.transaction.size'] != 0:
        connection.commit()


# ------------------------------------------------------------------------------
# Main routine.
# ------------------------------------------------------------------------------

def main():
    logging.basicConfig(format='%(levelname)s:%(message)s', level=logging.INFO)
    logging.info('Start OraBench.py')

    locale.setlocale(locale.LC_ALL, 'de_DE')

    run_benchmark()

    logging.info('End   OraBench.py')


# ------------------------------------------------------------------------------
# Performing the benchmark run.
# ------------------------------------------------------------------------------

def run_benchmark():
    config = get_config()

    measurement_data_result_file = create_result_measuring_point_start_benchmark(config)

    measurement_data = measurement_data_result_file[0]
    result_file = measurement_data_result_file[1]

    bulk_data_partitions = get_bulk_data_partitions(config)

    connections_cursors = create_database_objects(config)

    connections = connections_cursors[0]
    cursors = connections_cursors[1]

    for trial_number in range(1, config['benchmark.trials']):
        run_trial(config, connections, cursors, bulk_data_partitions, measurement_data, result_file, trial_number)

    for cursor in cursors:
        cursor.close()

    for connection in connections:
        connection.close()

    create_result_measuring_point_end(config, result_file, measurement_data, 'benchmark')


# ------------------------------------------------------------------------------
# Performing the insert operations.
# ------------------------------------------------------------------------------

def run_insert(config, connections, cursors, bulk_data_partitions, result_file, measurement_data, trial_number):
    measurement_data = create_result_measuring_point_start(measurement_data, 'query')

    threads = list()

    for partition_key in range(0, config['benchmark.number.partitions'] - 1):
        if config['benchmark.core.multiplier'] == 0:
            insert(config, connections[partition_key], cursors[partition_key], bulk_data_partitions[partition_key])
        else:
            thread = threading.Thread(target=insert, args=(config, connections[partition_key], cursors[partition_key], bulk_data_partitions[partition_key],))
            threads.append(thread)
            thread.start()

    if config['benchmark.core.multiplier'] > 0:
        for thread in threads:
            thread.join()

    create_result_measuring_point_end(config, result_file, measurement_data, 'query', trial_number, config['sql.insert'], 'insert')

    return measurement_data


# ------------------------------------------------------------------------------
# Performing the select operations.
# ------------------------------------------------------------------------------

def run_select(config, cursors, bulk_data_partitions, result_file, measurement_data, trial_number):
    measurement_data = create_result_measuring_point_start(measurement_data, 'query')

    threads = list()

    for partition_key in range(0, config['benchmark.number.partitions'] - 1):
        if config['benchmark.core.multiplier'] == 0:
            select(cursors[partition_key], bulk_data_partitions[partition_key], partition_key, config['sql.select'])
        else:
            thread = threading.Thread(target=select, args=(cursors[partition_key], bulk_data_partitions[partition_key], partition_key, config['sql.select'],))
            threads.append(thread)
            thread.start()

    if config['benchmark.core.multiplier'] > 0:
        for thread in threads:
            thread.join()

    create_result_measuring_point_end(config, result_file, measurement_data, 'query', trial_number, config['sql.select'], 'select')

    return measurement_data


# ------------------------------------------------------------------------------
# Performing one trial.
# ------------------------------------------------------------------------------

def run_trial(config, connections, cursors, bulk_data_partitions, measurement_data, result_file, trial_number):
    measurement_data = create_result_measuring_point_start(measurement_data, 'trial')

    logging.info('Start trial no. ' + str(trial_number))

    try:
        cursors[0].execute(config['sql.create'])
        logging.debug('last DDL statement=' + config['sql.create'])
    except cx_Oracle.DatabaseError:
        cursors[0].execute(config['sql.drop'])
        cursors[0].execute(config['sql.create'])
        logging.debug('last DDL statement after DROP=' + config['sql.create'])

    run_insert(config, connections, cursors, bulk_data_partitions, result_file, measurement_data, trial_number)

    run_select(config, cursors, bulk_data_partitions, result_file, measurement_data, trial_number)

    cursors[0].execute(config['sql.drop'])
    logging.debug('last DDL statement=' + config['sql.drop'])

    create_result_measuring_point_end(config, result_file, measurement_data, 'trial', trial_number)

    return measurement_data


# ------------------------------------------------------------------------------
# Performing the select operations.
# ------------------------------------------------------------------------------

def select(cursor, bulk_size_partition, partition_key, sql_statement):
    count = 0

    cursor.execute(sql_statement + ' where partition_key = ' + str(partition_key))

    for _ in cursor:
        count += 1

    if count != len(bulk_size_partition):
        logging.error('Number rows: expected=' + str(len(bulk_size_partition)) + ' - found=' + str(count))
        sys.exit(1)


# ------------------------------------------------------------------------------
# Program start.
# ------------------------------------------------------------------------------

main()

File: src_elixir/lib/ora_bench/ora_bench.ex
defmodule OraBench do
  require Logger

  @benchmark_module "OraBench (Elixir #{System.version()})"

  @file_configuration_name  "priv/properties/ora_bench.properties"

  @moduledoc false

  # ----------------------------------------------------------------------------------------------
  # Writing the results.
  # ----------------------------------------------------------------------------------------------

  def create_result(
        action,
        benchmark_driver,
        config,
        measurement_data,
        result_file,
        sql_operation,
        sql_statement,
        start_date_time,
        trial_number
      ) do
    Logger.debug("Start ==========> action: #{action} <==========")

    end_date_time = DateTime.utc_now()

    duration_ns = DateTime.diff(end_date_time, start_date_time, :nanosecond)
    duration_ss = DateTime.diff(end_date_time, start_date_time, :second)

    measurement_data_end = case sql_operation do
      "insert" ->
        Map.put(
          measurement_data,
          :duration_insert_sum,
          measurement_data[:duration_insert_sum] + duration_ns
        )
      "select" ->
        Map.put(
          measurement_data,
          :duration_select_sum,
          measurement_data[:duration_select_sum] + duration_ns
        )
      _ -> measurement_data
    end
    #    IO.inspect(measurement_data_end, label: "measurement_data_end")

    :ok = IO.puts(
      result_file,
      [
        config["benchmark.id"],
        config["file.result.delimiter"],
        config["benchmark.comment"],
        config["file.result.delimiter"],
        config["benchmark.host.name"],
        config["file.result.delimiter"],
        config["benchmark.number.cores"],
        config["file.result.delimiter"],
        config["benchmark.os"],
        config["file.result.delimiter"],
        config["benchmark.user.name"],
        config["file.result.delimiter"],
        config["benchmark.database"],
        config["file.result.delimiter"],
        @benchmark_module,
        config["file.result.delimiter"],
        benchmark_driver,
        config["file.result.delimiter"],
        Integer.to_string(trial_number),
        config["file.result.delimiter"],
        sql_statement,
        config["file.result.delimiter"],
        config["benchmark.core.multiplier"],
        config["file.result.delimiter"],
        config["connection.fetch.size"],
        config["file.result.delimiter"],
        config["benchmark.transaction.size"],
        config["file.result.delimiter"],
        config["file.bulk.length"],
        config["file.result.delimiter"],
        config["file.bulk.size"],
        config["file.result.delimiter"],
        config["benchmark.batch.size"],
        config["file.result.delimiter"],
        action,
        config["file.result.delimiter"],
        String.replace(
          Calendar.ISO.datetime_to_string(
            start_date_time.year,
            start_date_time.month,
            start_date_time.day,
            start_date_time.hour,
            start_date_time.minute,
            start_date_time.second,
            start_date_time.microsecond,
            start_date_time.time_zone,
            start_date_time.zone_abbr,
            start_date_time.utc_offset,
            start_date_time.std_offset
          ),
          "Z",
          "000"
        ),
        config["file.result.delimiter"],
        String.replace(
          Calendar.ISO.datetime_to_string(
            end_date_time.year,
            end_date_time.month,
            end_date_time.day,
            end_date_time.hour,
            end_date_time.minute,
            end_date_time.second,
            end_date_time.microsecond,
            end_date_time.time_zone,
            end_date_time.zone_abbr,
            end_date_time.utc_offset,
            end_date_time.std_offset
          ),
          "Z",
          "000"
        ),
        config["file.result.delimiter"],
        Integer.to_string(duration_ss),
        config["file.result.delimiter"],
        Integer.to_string(round(duration_ns))
      ]
    )

    measurement_data_end
  end

  # ----------------------------------------------------------------------------------------------
  # Creating the result file.
  # ----------------------------------------------------------------------------------------------

  def create_result_file(config) do
    Logger.debug("Start ==========> <==========")

    result_file_name = Path.expand(
                         ~s(../#{config["file.result.name"]})
                       )
                       |> Path.absname

    case File.exists?(result_file_name) do
      true ->
        {:ok, file} = File.open(result_file_name, [:append])
        file
      _ ->
        {:ok, file} = File.open(result_file_name, [:write])
        :ok = IO.puts(
          file,
          String.replace(
            config["file.result.header"],
            ";",
            config["file.result.delimiter"]
          )
        )
        file
    end
  end

  # ----------------------------------------------------------------------------------------------
  # Recording the results of the benchmark - end processing.
  # ----------------------------------------------------------------------------------------------

  def create_result_measuring_point_end(
        action,
        benchmark_driver,
        config,
        measurement_data,
        result_file,
        sql_operation,
        sql_statement,
        trial_number
      ) do
    Logger.debug("Start ==========> action: #{action} <==========")

    measurement_data_end = case action do
      "query" ->
        create_result(
          action,
          benchmark_driver,
          config,
          measurement_data,
          result_file,
          sql_operation,
          sql_statement,
          measurement_data[:last_query],
          trial_number
        )
      "trial" ->
        create_result(
          action,
          benchmark_driver,
          config,
          measurement_data,
          result_file,
          sql_operation,
          sql_statement,
          measurement_data[:last_trial],
          trial_number
        )
      "benchmark" ->
        create_result(
          action,
          benchmark_driver,
          config,
          measurement_data,
          result_file,
          sql_operation,
          sql_statement,
          measurement_data[:last_benchmark],
          trial_number
        )
        File.close(result_file)
      _ ->
        raise(
          "[Error in create_result_measuring_point_end] unknown action='#{
            action
          }'"
        )
    end
    #   IO.inspect(measurement_data_end, label: "measurement_data_end")

    measurement_data_end
  end

  # ----------------------------------------------------------------------------------------------
  # Recording the results of the benchmark - start processing.
  # ----------------------------------------------------------------------------------------------

  def create_result_measuring_point_start(action, measurement_data) do
    Logger.debug("Start ==========> action: #{action} <==========")

    case action do
      "query" ->
        Map.put(measurement_data, :last_query, DateTime.utc_now())
      "trial" ->
        Map.put(measurement_data, :last_trial, DateTime.utc_now())
      _ ->
        raise(
          "[Error in create_result_measuring_point_start] unknown action='#{
            action
          }'"
        )
    end
  end

  # ----------------------------------------------------------------------------------------------
  # Database driver configuration.
  # ----------------------------------------------------------------------------------------------

  def driver_config(config)  do
    Logger.debug("Start ==========> <==========")

    [
      database: config["connection.service"],
      hostname: config["connection.host"],
#      parameters: [
#        autocommit: false,
#        fetch: String.to_integer(config["connection.fetch.size"])
#      ],
      password: config["connection.password"],
#      pool_size: case config["benchmark.core.multiplier"] do
#        "0" -> String.to_integer(config["benchmark.number.partitions"])
#        _ -> 1
#      end,
      port: String.to_integer(config["connection.port"]),
      timeout: :infinity,
      username: config["connection.user"]
    ]
  end

  # ----------------------------------------------------------------------------------------------
  # Loading the bulk file into memory.
  # ----------------------------------------------------------------------------------------------

  def get_bulk_data_partitions(config) do
    Logger.debug("Start ==========> <==========")

    bulk_data_partitions = Path.expand(~s(../#{config["file.bulk.name"]}))
                           |> Path.absname
                           |> File.stream!()
                           |> Stream.map(&(String.replace(&1, "\n", "")))
                           |> Enum.reduce(
                                %{},
                                fn (line, bulk_data_partitions) ->
                                  case line == config["file.bulk.header"] do
                                    true ->
                                      bulk_data_partitions
                                    _ ->
                                      case String.split(line, config["file.bulk.delimiter"]) do
                                        [key, data] ->
                                          partition_key = rem(
                                            (String.at(key, 0)
                                             |> String.to_charlist()
                                             |> hd) * 256 + (String.at(key, 1)
                                                             |> String.to_charlist()
                                                             |> hd),
                                            String.to_integer(config["benchmark.number.partitions"])
                                          )
                                          case Map.has_key?(
                                                 bulk_data_partitions,
                                                 partition_key
                                               ) do
                                            true ->
                                              Map.put(
                                                bulk_data_partitions,
                                                partition_key,
                                                [
                                                  {key, data} | bulk_data_partitions[partition_key]
                                                ]
                                              )
                                            _ ->
                                              Map.put(
                                                bulk_data_partitions,
                                                partition_key,
                                                [{key, data}]
                                              )
                                          end
                                        _ ->
                                          bulk_data_partitions
                                      end
                                  end
                                end
                              )


    Logger.info("Start Distribution of the data in the partitions")

    Enum.each(
      Map.keys(bulk_data_partitions),
      fn key ->
        Logger.info(
          "Partition p#{
            key
            |> Integer.to_string
            |> String.pad_leading(5, "0")
          } contains #{
            length(bulk_data_partitions[key])
            |> Integer.to_string
            |> String.pad_leading(9, " ")
          } rows"
        )
      end
    )

    Logger.info("End   Distribution of the data in the partitions")

    bulk_data_partitions
  end

  # ----------------------------------------------------------------------------------------------
  # Loading the configuration parameters into memory.
  # ----------------------------------------------------------------------------------------------

  def get_config do
    Logger.debug("Start ==========> <==========")

    config = Path.expand("../#{@file_configuration_name}")
             |> Path.absname
             |> File.stream!()
             |> Stream.map(&(String.replace(&1, "\n", "")))
             |> Enum.reduce(
                  %{},
                  fn (line, config) ->
                    case String.split(line, "=", [parts: 2]) do
                      [key, value] -> Map.put_new(config, key, value)
                      _ -> config
                    end
                  end
                )

    case config["file.result.delimiter"] do
      "\\t" -> Map.put(config, "file.result.delimiter", <<"\t">>)
      _ -> config
    end
  end
end
